Sources/SocksCore/TCPSocket.swift

```diff
@@ -203,12 +203,15 @@ public class TCPInternetSocket: InternetSocket, TCPSocket, TCPReadableSocket, TC
     public func close() throws {
         if closed { return }
         stopWatching()
+        closed = true
         if socket_close(self.descriptor) != 0 {
+            if errno != EBADF {
```

Comment:
`closed` var has to be set to `true` before trying to close the socket. Otherwise application checking `closed` var to send or receive data on the socket can be in inconsistent state. I've detected this error running benchmark with [MongoKitten](https://github.com/OpenKitten/Statistics). Settings `closed` earlier prevent communication on closing socket. It's not clear for me how to handle the other error case. Obviously  EBADF is clearly closed :-)  so `descriptor` has to be set to -1.